man/ipsvd-cdb.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "ipsvd\-cdb" "8" 
.SH NAME
ipsvd\-cdb \- create constant database from ipsvd instructions directory
.SH SYNOPSIS
\f[B]ipsvd\-cdb\f[R] \f[I]cdb\f[R] \f[I]cdb.tmp\f[R] \f[I]dir\f[R]
.SH DESCRIPTION
\f[B]ipsvd\-cdb\f[R] scans the directory \f[I]dir\f[R], and compiles all
relevant information for an internet protocol service daemon,
\f[B]ipsvd\f[R](7), into the newly created constant database
\f[I]cdb.tmp\f[R].
If \f[I]cdb.tmp\f[R] exists, it will be destroyed.
\f[B]ipsvd\-cdb\f[R] then renames \f[I]cdb.tmp\f[R] to \f[I]cdb\f[R].
If \f[I]cdb\f[R] exists, it will be destroyed.
.PP
\f[B]ipsvd\-cdb\f[R] skips filenames starting with two dots.
.PP
If \f[B]ipsvd\-cdb\f[R] has trouble building the constant database, it
exits with an error message, and leaves \f[I]cdb\f[R] unchanged.
.SH SEE ALSO
ipsvd(7), tcpsvd(8), udpsvd(8), ipsvd\-instruct(5)
.PP
https://smarden.org/ipsvd/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * doc/examples.html: use chpst instead of softlimit.
  * doc/install.html: how to install as non-root user.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

ipsvd 1.1.1
Sun, 02 Nov 2025 11:09:29 +0100
  * alloc.c, alloc_re.c, buffer_0.c, buffer_get.c, buffer_put.c, byte_chr.c,
    byte_copy.c, byte_cr.c, byte_diff.c, byte_rchr.c, byte_zero.c, error.c,
    error.h, print-cc.sh, print-ld.sh, select.h2, strerr_die.c, taia_sub.c,
    trysgprm.c, wait.h, wait_nohang.c, wait_pid.c: sync daemontools library
    fork with runit source.
  * tcpsvd.c, udpsvd.c: use parameter name "unused" in sighandler functions.
  * hasflock.h1: add #include <unistd.h>.
  * doc/*.html: !DOCTYPE.
  * doc/index.html: Related links: remove links to distributions, repair
    broken links.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*: switch all smarden.org http links to https.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-udpsvd.c, ipsvd_phcc.c, pathexec_env.c, select.h1,
    select.h2, sig.c, sig.h, sig_catch.c, socket_bind.c, socket_conn.c,
    tcpsvd.c, tcpsvd.dist, udpsvd.c, udpsvd.dist, wait.h: fix build with
    gcc15: in C23 f() means f(void), consequently signal handler functions
    now are f(int), alloc_free() and byte_zero() parameters need casts,
    buffer_unixwrite() removes "const" from parameter.
  * src/Makefile: target check depends on target default now.
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.
  * doc/install.html: put back installation step removed by mistake.

ipsvd 1.1.0
Wed, 16 Oct 2024 20:36:33 +0000
  * doc/benefits.html, doc/examples.html. doc/index.html,
    doc/install.html: remove references to sslio, sslsvd.
  * man/ipsvd-cdb.8, man/ipsvd-instruct.5, man/ipsvd.7, man/tcpsvd.8,
    man/udpsvd.8: remove references to sslio.8, sslsvd.8.
  * man/sslio.8, man/sslsvd.8: remove; obsolete.
  * tcpsvd.c, tcpsvd.dist: remove sslsvd code; obsolete.
  * *ssl*: remove; obsolete.
  * Makefile, TARGETS, check-dist, package/compile, package/upgrade:
    remove support for matrixssl; no longer build sslio, sslsvd programs.
  * doc/upgrade.html: typo.
  * tcpsvd.c, udpsvd.c: print proper group id when verbose.
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * debian/: remove; obsolete.
  * cdb_make.c, pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c,
    sgetopt.c, sgetopt.h, subgetopt.c, subgetopt.h, tcpsvd.c, trycpp.c,
    tryflock.c, trypoll.c, trysgact.c, trysgprm.c, tryulong64.c,
    trywaitp.c, udpsvd.c, x86cpuid.c: incorporate lib changes from
    runit-2.2.0 and adapt ipsvd programs to properly build with modern
    toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * ssl_io.c, tcpsvd.c, udpsvd.c: replace prot_uid() with setuid().

ipsvd 1.0.0
Sun, 15 Jun 2008 17:45:26 +0000

ipsvd 0.14.0
Sat, 19 Jan 2008 00:17:20 +0000
  * uidgid.c, uidgid.h: support numerical uid, gids through :uid:gid...
    (taken from the runit package).
  * tcpsvd.c, udpsvd.c, sslio.c, ssl_io.c: support numerical uid, gids,
    and colon-separated list of groups through :uid:gid... with the -u,
    -U arguments (thx Nikola Vladov).
  * man/tcpsvd.8, man/udpsvd.8, man/sslsvd.8, man/sslio.8, man/ipsvd.7:
    adapt.

ipsvd 0.13.0
Sun, 26 Aug 2007 11:04:21 +0000
  * ipsvd_check.c, man/ipsvd-instruct.5: lines starting with # are
    comments, ignore them completely.
  * ssl_io.c, man/sslio.8: support env variable SSLIO_BAD_CERTIFICATE to
    optionally accept 'bad certificate's (thx Nikola Vladov).
  * ssl_io.c, man/sslio.8, man/sslsvd.8: support env variable
    SSL_HANDSHAKE_TIMEOUT to disconnect a client if the ssl handshake
    isn't completed within some number of seconds (thx Nikola Vladov);
    add ENVIRONMENT section to man/sslsvd.8.
  * Makefile: build with matrixssl-1-8-3-open.

ipsvd 0.12.1
Sat, 04 Feb 2006 19:14:23 +0000
  * Makefile: fix build failure with local matrixssl library (thx Charlie
    Brady).

ipsvd 0.12.0
Fri, 03 Feb 2006 20:26:27 +0000
  * ssl_io.c, ssl_io.h: new; split off most code from sslio.c.
  * sslio.c: main() function only, uses ssl_io() from ssl_io.
  * tcpsvd.c: be sslsvd using ssl_io() #ifdef SSLSVD.
  * sslsvd.c: new; #define SSLSVD, #include tcpsvd.c.
  * man/sslsvd.8: new; man page for sslsvd(8).
  * man/ipsvd-cdb.8, man/ipsvd-instruct.5, man/ipsvd.7, man/sslio.8,
    man/tcpsvd.8, man/udpsvd.8: add SEE ALSO sslsvd(8).
  * sslsvd.check, sslsvd.dist: selftest for sslsvd not yet implemented.
  * doc/index.html: mention sslsvd.

ipsvd 0.11.1
Tue, 18 Oct 2005 19:02:18 +0000
  * sslio.check: tweak sslio program check (fixes testing failure on
    Debian m68k)
  * ipsvd_log.c: minor
  * Makefile, doc/install.html: adapt to matrixssl-1-7-1-open.tar.gz.

ipsvd 0.11.0
Sun, 20 Feb 2005 16:13:38 +0000
  * man/tcpsvd.8: typo (thx Bastian Kleineidam).
  * sslio.check, sslio.dist, sslio.pem: new; ssl program check.
  * Makefile: don't limit sslio program to linux systype; run ssl program
    check; minor
  * package/compile, package/upgrade: adapt.
  * doc/benefits.html, doc/index.html, doc/install.html: sslio works on
    MacOSX; adapt.

ipsvd 0.10.1
Sun, 16 Jan 2005 10:34:19 +0000
  * sslerror_str.c: fix build failure due to broken header include (thx
    Charlie Brady)
  * Makefile: target sslerror_str.o depends on matrixSsl.h.
  * sslio.c: minor.

ipsvd 0.10.0
Sat, 15 Jan 2005 11:47:55 +0000
  * doc/install.html: typo.
  * sslerror_str.c, sslerror_str.h: new; ssl error strings.

src/ipsvd-cdb.c
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <unistd.h>
#include <dirent.h>
#include "ipsvd_check.h"
#include "sgetopt.h"
#include "error.h"
#include "open.h"
#include "lock.h"
#include "strerr.h"
#include "stralloc.h"
#include "direntry.h"
#include "cdb.h"
#include "cdb_make.h"
#include "str.h"

#define USAGE " cdb cdb.tmp dir"
#define VERSION "$Id$"
#define FATAL "ipsvd-cdb: fatal: "
#define WARNING "ipsvd-cdb: warning: "

const char *progname;
char *instdir;
char *cdbfn;
char *tmpfn;

struct cdb_make c;
struct cdb cdb;
int fdcdb;
int fdtmp;
static stralloc sa ={0};
static stralloc tmp ={0};

void usage() { strerr_die4x(111, "usage: ", progname, USAGE, "\n"); }
void die_nomem() { strerr_die2x(111, FATAL, "out of memory."); }
void fatal(char *m0) { strerr_die3sys(111, FATAL, m0, ": "); }
void fatal2(char *m0, char *m1) {
  strerr_die5sys(111, FATAL, m0, ": ", m1, ": ");
}
void warn(char *m0, char *m1) { strerr_warn4(WARNING, m0, ": ", m1, 0); }

int main(int argc, char **argv) {
  int mydir;
  DIR *dir;
  direntry *d;
  struct stat s;
  int ac;
  int i;

  progname =*argv++;

  if (! argv || ! *argv) usage();
  cdbfn =*argv++;
  if (! argv || ! *argv) usage();
  tmpfn =*argv++;
  if (! argv || ! *argv) usage();
  instdir =*argv;

  if ((mydir =open_read(".")) == -1)
    fatal("unable to open current directory");

  /* open cdb.tmp */
  if ((fdtmp =open_trunc(tmpfn)) == -1) fatal2("unable to create", tmpfn);
  if (cdb_make_start(&c, fdtmp) == -1) fatal2("unable to create", tmpfn);

  if (chdir(instdir) == -1) fatal2("unable to change dir", instdir);
  if (! (dir =opendir("."))) fatal2("unable to open dir", instdir);
  errno =0;
  while ((d =readdir(dir))) {
    if (d->d_name[0] == '.') {
      if (d->d_name[1] == 0) continue;
      if (d->d_name[1] == '.') continue;
    }
    if (stat(d->d_name, &s) == -1) {
      warn("unable to stat", d->d_name);
      errno =0;
      continue;
    }
    sa.len =0;
    ac =ipsvd_check(0, &sa, &tmp, ".", 0, d->d_name, 0);
    if (ac == -1) fatal2("unable to read instruction", d->d_name);
    if (ac == IPSVD_ERR) fatal2("unable to read", "."); /* impossible? */

    switch(ac) {
    case IPSVD_EXEC:
      sa.s[sa.len -1] ='X';
      if (cdb_make_add(&c, d->d_name, str_len(d->d_name), sa.s, sa.len) == -1)
        fatal2("unable to add entry", instdir);
      continue;
    case IPSVD_DENY:
      if (! sa.len) {
        if (cdb_make_add(&c, d->d_name, str_len(d->d_name), "D", 1) == -1)
          fatal2("unable to add entry", instdir);
        continue;
      }
    case IPSVD_INSTRUCT:
      for (i =0; i < sa.len; i++) if (sa.s[i] == '\n') sa.s[i] =0;
      sa.s[sa.len -1] ='I';
      if (cdb_make_add(&c, d->d_name, str_len(d->d_name), sa.s, sa.len) == -1)
        fatal2("unable to add entry", instdir);
      continue;
    }
    warn("ignore", d->d_name);
  }
  /* catch errno? */
  closedir(dir);
  if (cdb_make_finish(&c) == -1) fatal2("unable to write cdb", tmpfn);
  if (fsync(fdtmp) == -1) fatal2("unable to write cdb", tmpfn);
  close(fdtmp);
  if (fchdir(mydir) == -1) fatal("unable to change to previous directory");
  close(mydir);
  if (rename(tmpfn, cdbfn) == -1) fatal2("unable to replace", cdbfn);
  _exit(0);
}
